saleor/plugins/openid_connect/utils.py

```diff
@@ -0,0 +1,577 @@
+        logger.warning("Multiple users returned for single OIDC sub ID")
+        user, _ = User.objects.get_or_create(
+            email=user_email,
+            defaults=defaults_create,
+        )
+
+    if not user.is_active:  # it is true only if we fetch disabled user.
+        raise AuthenticationError("Unable to log in.")
+
+    _update_user_details(
+        user=user,
+        oidc_key=oidc_metadata_key,
+        user_email=user_email,
+        sub=sub,  # type: ignore
+        last_login=last_login,
+    )
+
+    return user
+
+
+def _update_user_details(
+    user: User, oidc_key: str, user_email: str, sub: str, last_login: Optional[int]
+):
+    user_sub = user.get_value_from_private_metadata(oidc_key)
+    fields_to_save = []
+    if user_sub != sub:
+        user.store_value_in_private_metadata({oidc_key: sub})
+        fields_to_save.append("private_metadata")
+
+    if user.email != user_email:
+        if User.objects.filter(email=user_email).exists():
+            logger.warning(
+                "Unable to update user email as the new one already exists in DB",
+                extra={"oidc_key": oidc_key},
+            )
+            return
+        user.email = user_email
+        fields_to_save.append("email")
+    if last_login:
+        if not user.last_login or user.last_login.timestamp() < last_login:
+            login_time = make_aware(datetime.fromtimestamp(last_login))
+            user.last_login = login_time
+            fields_to_save.append("last_login")
+
+    if fields_to_save:
+        user.save(update_fields=fields_to_save)
+
+
+def get_user_from_token(claims: CodeIDToken) -> User:
+    user_email = claims.get("email")
+    if not user_email:
+        raise AuthenticationError("Missing user's email.")
+    user = User.objects.filter(email=user_email, is_active=True).first()
+    if not user:
+        raise AuthenticationError("User does not exist.")
+    return user
+
+
+def is_owner_of_token_valid(token: str, owner: str) -> bool:
+    try:
+        payload = jwt_decode(token, verify_expiration=False)
+        return payload.get(JWT_OWNER_FIELD, "") == owner
+    except Exception:
+        return False
+
+
+def create_tokens_from_oauth_payload(
+    token_data: dict,
+    user: User,
+    claims: CodeIDToken,
+    permissions: Optional[List[str]],
+    owner: str,
+):
+    refresh_token = token_data.get("refresh_token")
+    access_token = token_data.get("access_token", "")
+
+    tokens = {
+        "token": create_jwt_token(claims, user, access_token, permissions, owner),
+    }
+    if refresh_token:
+        csrf_token = _get_new_csrf_token()
+        tokens["refresh_token"] = create_jwt_refresh_token(
+            user, refresh_token, csrf_token, owner
+        )
+        tokens["csrf_token"] = csrf_token
+    return tokens
+
+
+def validate_refresh_token(refresh_token, data):
+    csrf_token = data.get("csrfToken")
+    if not refresh_token:
+        raise ValidationError(
```

Comment:
IMO this shouldn't return an ValidationError but some other class, but I guess we don't have time for this now.